server-scripts/simple-incremental-diff-server.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+#
+# Incremental Diff Server
+#
+# For POST requests:
+#   Reads PullRequest event data from github and extracts the HEAD SHA from it.
+#   The SHA is then stored at the <basedir> location using the pattern
+#   <owner>/<repo>/<pr-number>
+#
+#   Only runs for PullRequest actions "synchronize" (when someone pushed new
+#   commits) and "opened" (when a new PR has been opened)
+#
+# For GET requests:
+#   Serves the files for <basedir>
+
+import os
+import sys
+import time
+import json
+import http.server
+import shutil
+
+import hmac
+from hashlib import sha1
+
+# Listening Port
+PORT = 8000
+# Github Secret, optional
+SECRET = ""
+# Base directory that will be used as root
+BASEDIR = "hashes/"
+
+class Handler(http.server.BaseHTTPRequestHandler):
+
+    def checkSecret(self, data):
+        header_sig = self.headers.get('X-Hub-Signature')
+
+        if header_sig is None:
+            print("Error: No signature found!")
+            return False
+
+        sha_name, signature = header_sig.split('=')
+        if sha_name != 'sha1':
+            print("Error: Unexpected hash algorithm")
+            return False
+
+        # HMAC requires the key to be bytes, but data is string
+        mac = hmac.new(bytearray(SECRET, 'utf8'), msg=data, digestmod=sha1)
+
+        # What compare_digest provides is protection against timing
+        # attacks; we can live without this protection for a web-based
+        # application
+        if not hmac.compare_digest(str(mac.hexdigest()), str(signature)):
+            print("Error: digest mismatch")
+            return False
+
+        return True
+
+    def do_POST(self):
+        self.send_response(200)
+        self.send_header("Content-Type", "text/html")
+        self.end_headers()
+
+        data = self.rfile.read(int(self.headers.get("Content-Length")))
+
+        if len(SECRET) > 0:
+            if self.checkSecret(data) == False:
+                return
+
+        basedir = BASEDIR
+
+        parsed = json.loads(data.decode("utf-8"))
+
+        if 'action' not in parsed:
+            print("No 'action' in payload, skipping...")
+            return
+
+        if parsed["action"] not in ("synchronize", "opened"):
+            print("Action not one of 'synchronize' or 'opened', skipping...")
+            return
+
+        sha = parsed["pull_request"]["head"]["sha"]
+        prid = parsed["number"]
+        repo = parsed["pull_request"]["base"]["repo"]["full_name"]
+
+        target_path = basedir + "/" + repo
+
+        if os.path.isdir(target_path) == False:
+            print("Creating ", target_path)
+            os.makedirs(target_path)
+
+        fname = "{}/{}".format(target_path, prid)
+        data = "{};{}\n".format(sha, int(time.time()))
+
+        print("Writing {!r} to {!r}...".format(data, fname))
+
+        with open(fname, "a") as f:
+            f.write(data)
```

Comment:
Now you are not using the external webhooks server at all? My suggestion was to just remove nginx and replace it with `python3 -m http.server`, but leave the full webhooks server too (just cloning one project seemed like a reasonable step. But if this works, it is in fact much more compact for users that just want to try it out. But it would be good to put (in the Wiki maybe) more recipes of how to use this, including that webhooks server (optionally with nginx) and maybe the Jenkins setup.